veps/sig-compute/hypervisor-feature-discovery.md

```diff
@@ -0,0 +1,373 @@
+- **Type-safe**: Constants prevent typos, enable IDE autocomplete
+- **Rich error messages**: Automatic user-friendly messages with doc links
+- **Automatic test filtering**: Tests skip only when explicitly unsupported
+
+## API Examples
+
+### Querying Capabilities
+
+```go
+// Check if a capability has constraints
+cap := capabilities.Get("kvm", "amd64", capabilities.CapGraphicsVGA)
+if cap.Level == capabilities.Unsupported {
+    // Explicitly blocked on this platform
+    return fmt.Errorf("%s: %s", capabilities.CapGraphicsVGA, cap.Message)
+}
+if cap.Level == capabilities.Experimental && !featureGateEnabled(cap.GatedBy) {
+    return fmt.Errorf("%s requires feature gate %s", capabilities.CapGraphicsVGA, cap.GatedBy)
+}
+// Unregistered: allow (normal code paths handle it)
+
+// Get all registered capabilities for a platform (only those explicitly declared)
+caps := capabilities.GetAll("kvm", "amd64")
+for key, cap := range caps {
+    log.Printf("%s: %s (since %s)", key, cap.Level, cap.Since)
+}
+```
+
+### Error Message Example
+
+```
+VGA graphics not supported on ARM64 architecture
+Capability: graphics.vga
+See: https://kubevirt.io/user-guide/graphics#compatibility
+```
+
+## Undesired alternatives
+
+1. **Status quo** – Continue scattered architecture-specific validation files. This increases maintenance burden and makes adding hypervisors difficult.
+2. **Per-hypervisor capability files** – Create `pkg/capabilities/hypervisor/kvm/amd64.go` etc. Rejected due to file sprawl and duplication.
+3. **Runtime-only discovery** – Query libvirt capabilities at runtime instead of static registry. Rejected because it requires running nodes and can't be used for validation or test planning.
+
+## Scalability
+
+- Capability lookups are constant-time map operations.
+- Registry is populated once during init, no runtime registration overhead.
+- Memory footprint scales linearly with number of hypervisor+arch+capability combinations (expected to be small).
+
+## Update/Rollback Compatibility
+
+- The capability registry is additive—coexists with existing validation logic during migration.
+- Rolling back removes capability-based validation but doesn't break existing functionality.
+
+## Functional Testing Approach
+
+- Unit tests for capability registry ensuring correct lookup precedence and fallback behavior.
+- Integration tests verifying validation webhooks reject unsupported features with correct error messages.
+- Integration tests confirming test decorators correctly skip tests on unsupported platforms.
+- Minimal end-to-end tests covering a happy path and a negative path to demonstrate the wiring; the bulk of the coverage lives in unit and integration tests that exercise resolution and webhook behavior.
+
+## Implementation History
+
+- 2025-Oct-7: Initial VEP #95 draft.
+- 2025-Oct-31: Initial drafet of VEP #95.1 (Extracted feature discovery from VEP #95).
+
+## Graduation Requirements
+
+### Alpha
```

Comment:
Can we specify in the different graduation stages what functionality is available and can be used? I am unclear about that as the details here go too deep.

E.g. "Introduce the infrastructure and existing KVM capabilities, in order to use them when ....".

To clarify, the expectation is to see a functional solution which can be experimented with. If a new hypervisor cannot be introduced with this, we need something else, like an architecture or something that is based on FG enablement (although, those are not init time based, but init runtime based).